sycl/test/extensions/bindless_images_SPIRV_inst.cpp

```diff
@@ -0,0 +1,125 @@
+// `7` here represents 'TypeVoid`. Must be `TypeVoid` as the type of the image
+// is not known at compile time. The last operand is the access qualifier. With
+// 0 read only and 1 write only.
+// Arguments: TypeImage, Result, Sampled Type, Dim, Depth, Arrayed, MS, Sampled
+// and Image Format.
+// CHECK: TypeImage [[IMAGETYPEREAD:[0-9]+]] 7 0 0 0 0 0 0 0
+// CHECK: TypeImage [[IMAGETYPEWRITE:[0-9]+]] 7 0 0 0 0 0 0 1
+// Generate `Result Type` for samplers
+// Arguments: Result
+// CHECK: TypeSampler [[SAMPLERTYPE:[0-9]+]]
+// Generate `Result Type` for sampled images
+// Arguments: Result, Image Type
+// CHECK: TypeSampledImage [[SAMPLEDIMAGETYPE:[0-9]+]] [[IMAGETYPEREAD]]
+
+// Convert handles to SPIR-V images, samplers and sampled images
+// Arguments: Result Type, Result, Handle
+// CHECK: ConvertHandleToImageINTEL [[IMAGETYPEREAD]] {{[0-9]+}} [[IMAGEID]]
+// CHECK: ConvertHandleToImageINTEL [[IMAGETYPEWRITE]] {{[0-9]+}} [[IMAGETWOID]]
+// CHECK: ConvertHandleToSamplerINTEL [[SAMPLERTYPE]] {{[0-9]+}} [[SAMPLERID]]
+// CHECK: ConvertHandleToSampledImageINTEL [[SAMPLEDIMAGETYPE]] {{[0-9]+}} [[SAMPLEDIMAGEID]]
+
+#include <sycl/detail/image_ocl_types.hpp>
+
+#ifdef __SYCL_DEVICE_ONLY__
+template <int NDims>
+using OCLImageTyRead =
+    typename sycl::detail::opencl_image_type<NDims, sycl::access::mode::read,
+                                             sycl::access::target::image>::type;
+
+template <int NDims>
+using OCLImageTyWrite =
+    typename sycl::detail::opencl_image_type<NDims, sycl::access::mode::write,
+                                             sycl::access::target::image>::type;
+
+template <int NDims>
+using OCLSampledImageTy = typename sycl::detail::sampled_opencl_image_type<
+    OCLImageTyRead<NDims>>::type;
+#endif
+
+template <typename ReturnT> ReturnT handleToImage(const uint64_t &imageHandle) {
+#ifdef __SYCL_DEVICE_ONLY__
+#if defined(__SPIR__)
+  return __spirv_ConvertHandleToImageINTEL<ReturnT>(imageHandle);
+#endif
+#else
+  assert(false); // Bindless images not yet implemented on host.
+#endif
+}
+
+template <typename ReturnT>
+ReturnT handleToSampler(const uint64_t &samplerHandle) {
+#ifdef __SYCL_DEVICE_ONLY__
+#if defined(__SPIR__)
+  return __spirv_ConvertHandleToSamplerINTEL<ReturnT>(samplerHandle);
+#endif
+#else
+  assert(false); // Bindless images not yet implemented on host.
+#endif
+}
+
+template <typename ReturnT>
+ReturnT handleToSampledImage(const uint64_t &sampledImageHandle) {
+#ifdef __SYCL_DEVICE_ONLY__
+#if defined(__SPIR__)
+  return __spirv_ConvertHandleToSampledImageINTEL<ReturnT>(sampledImageHandle);
+#endif
+#else
+  assert(false); // Bindless images not yet implemented on host.
+#endif
+}
+
+class image_addition;
+
+int main() {
+
+  sycl::device dev;
+  sycl::queue q(dev);
+  auto ctxt = q.get_context();
+  try {
+    q.submit([&](sycl::handler &cgh) {
+      cgh.parallel_for<image_addition>(1, [=](sycl::id<1> id) {
+#ifdef __SYCL_DEVICE_ONLY__
```

Comment:
That can become troublesome if customers have to structure their code like this. In the worst case it can lead to different lambda captures between host/device.

Reply:
Customers do not need to structure their code like this. I was just intending to invoke the new `__spirv_` builtins manually to verify they worked. The guards are needed as opencl image types do not compile on host.

There is the other test, `sycl/test/extensions/bindless_images.cpp` which does also check that the SPIR-V instructions are emitted correctly, but it does not check that `__spirv_ConvertHandleToSamplerINTEL` works correctly as it is currently not used in the SYCL bindless extension. So I wanted to keep this to ensure all of the new builtins were covered.